modules/pam_loginuid/pam_loginuid.8
.\"     Title: pam_loginuid
.\"    Author: 
.\" Generator: DocBook XSL Stylesheets vsnapshot_2006\-08\-24_0226 <http://docbook.sf.net/>
.\"      Date: 09/06/2006
.\"    Manual: Linux\-PAM Manual
.\"    Source: Linux\-PAM Manual
.\"
.TH "PAM_LOGINUID" "8" "09/06/2006" "Linux\-PAM Manual" "Linux\-PAM Manual"
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.SH "NAME"
pam_loginuid \- Record user's login uid to the process attribute
.SH "SYNOPSIS"
.HP 16
\fBpam_loginuid.so\fR [require_auditd]
.SH "DESCRIPTION"
.PP
The pam_loginuid module sets the loginuid process attribute for the process that was authenticated. This is necessary for applications to be correctly audited. This PAM module should only be used for entry point applications like: login, sshd, gdm, vsftpd, crond and atd. There are probably other entry point applications besides these. You should not use it for applications like sudo or su as that defeats the purpose by changing the loginuid to the account they just switched to.
.SH "OPTIONS"
.TP 3n
\fBrequire_auditd\fR
This option, when given, will cause this module to query the audit daemon status and deny logins if it is not running.
.SH "MODULE SERVICES PROVIDED"
.PP
The
\fBsession\fR
service is supported.
.SH "RETURN VALUES"
.PP
.TP 3n
PAM_SESSION_ERR
An error occured during session management.
.SH "EXAMPLES"
.sp
.RS 3n
.nf
#%PAM\-1.0
auth       required     pam_unix.so
auth       required     pam_nologin.so
account    required     pam_unix.so
password   required     pam_unix.so
session    required     pam_unix.so
session    required     pam_loginuid.so
    
.fi
.RE
.SH "SEE ALSO"
.PP

\fBpam.conf\fR(5),
\fBpam.d\fR(8),
\fBpam\fR(8),
\fBauditctl\fR(8),
\fBauditd\fR(8)
.SH "AUTHOR"
.PP
pam_loginuid was written by Steve Grubb <sgrubb@redhat.com>


NEWS
Linux-PAM NEWS -- history of user-visible changes.


Release 0.99.6.3

* pam_loginuid: New PAM module.
* pam_access, pam_succeed_if: Support passwd and session services.

Release 0.99.6.2

* pam_lastlog: Don't refuse login if lastlog file got lost.
* pam_cracklib: Fix a user triggerable crash.
* documentation: Regenerate with fixed docbook stylesheet.


Release 0.99.6.1

* Fix bootstrapping problems.
* Bug fixes: pam_keyinit, pam_umask


Release 0.99.6.0

* pam_namespace: Code cleanup, add init script to tar archive.
* pam_succeed_if: Add support for service match.
* Add xtests (to run after installation).
* Documentation: Convert sgml guides to XML, unify documentation
  for PAM functions and modules.


Release 0.99.5.0

* pam_tally: Fix support for large UIDs
* Fixed all problems found by Coverity
* Add support for Intel C Compiler
* Add manual page for pam_mkhomedir, pam_umask, pam_filter,
  pam_issue, pam_ftp, pam_group, pam_lastlog, pam_listfile,
  pam_localuser, pam_mail, pam_motd, pam_nologin, pam_permit,
  pam_rootok, pam_securetty, pam_shells, pam_userdb, pam_warn,
  pam_time, pam_limits, pam_debug, pam_tally
* The libpam memory debug code was removed
* pam_keyinit: New module to initialise kernel session keyring.
* pam_namespace: New module to configure private namespace for a session.
* pam_rhosts: New module which replaces pam_rhosts_auth, now IPv6 capable.
* pam_rhosts_auth: This module is now deprecated.


Release 0.99.4.0

* Add test suite
* Fix building of static variants of libpam, libpamc and libpam_misc
* pam_listfile: Add support for password and session management
* pam_exec: New PAM module to execute arbitary commands
* Fix building of a static libpam including all PAM modules
* New/updated translations for: nl, pt, pl, fi, km, tr, uk, fr
* pam_access: Add network(address) / netmask and IPv6 support
* Add manual pages for pam_cracklib, pam_deny and pam_access
* pam_pwdb: This deprecated module was removed
* Manual pages: Major rewrite/cleanup


Release 0.99.3.0

* Fix NULL pointer checks in libpam.so
* pam_succeed_if, pam_group, pam_time: Support netgroup matching
* New translations for: nb, hu, fi, de, es, fr, it, ja, pt_BR, zh_CN, zh_TW
* Audit PAM calls if Linux Audit is available
* Compile upperLOWER and unix_chkpwd as PIE binaries


Release 0.99.2.1

* Fix install of PS, PDF, TXT and HTML files
* pam_mail: Update README
* Use %m consistent
* pam_modutil_getlogin: Fix parsing of PAM_TTY variable


Release 0.99.2.0

* Fix parsing of full path tty name in various modules
* pam_xauth: Look for xauth executable in multiple places
* pam_unix: Disable user check in unix_chkpwd only if real uid
  is 0 (CVE-2005-2977). Log failed password check attempt.
* pam_env: Support /etc/environment again, but don't treat it as
  error if it is missing.
* pam_userdb: Fix memory leak.


Release 0.99.1.0

* Use autoconf/automake/libtool
* Add gettext support
* Add translations for cs, de, es, fr, hu, it, ja, nb, pa, pt_BR,
  pt, zh_CN and zh_TW
* libpam: Remove pam_authenticate_secondary stub
* libpam: Add pam_prompt,pam_vprompt,pam_error,pam_verror,pam_info
  and pam_vinfo functions for use by modules as extension
* libpam: Add pam_syslog function for unified syslog messages from
  PAM modules
* libpam: Moved functions from pammodutil to libpam
* pam_umask: New module for setting umask from GECOS field, /etc/login.defs
  or /etc/default/login
* pam_echo: New PAM module for message output
* pam_userdb: Fix regression (crash when crypt param not specified)
* pam_limits: Fix regression from RLIMIT_NICE support (wrong limit
  values for other limits are applied)
* pam_access: Support for NULL tty - matches ALL and NONE keywords
* pam_lastlog: Enable log to wtmp by default.  Add "nowtmp" option
* pam_radius: This module was removed


modules/pam_loginuid/pam_loginuid.c
/* pam_loginuid.c --
 * Copyright 2005 Red Hat Inc., Durham, North Carolina.
 * All Rights Reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 * Authors:
 *   Steve Grubb <sgrubb@redhat.com>
 *
 * PAM module that sets the login uid introduced in kernel 2.6.11
 */

#include "config.h"
#include <stdio.h>
#include <stdarg.h>
#include <syslog.h>
#include <string.h>
#include <pwd.h>
#include <unistd.h>
#include <limits.h>
#include <errno.h>

#include <security/pam_modules.h>
#include <security/pam_ext.h>
#include <security/pam_modutil.h>

#include <fcntl.h>

#ifdef HAVE_LIBAUDIT
#include <libaudit.h>
#include <sys/select.h>
#include <errno.h>
#endif

/*
 * This function writes the loginuid to the /proc system. It returns
 * 0 on success and 1 on failure.
 */
static int set_loginuid(pam_handle_t *pamh, uid_t uid)
{
	int fd, count, rc = 0;
	char loginuid[24];

	count = snprintf(loginuid, sizeof(loginuid), "%d", uid);
	fd = open("/proc/self/loginuid", O_NOFOLLOW|O_WRONLY|O_TRUNC);
	if (fd < 0) {
		int loglevel = LOG_DEBUG;
		if (errno != ENOENT) {
			rc = 1;
			loglevel = LOG_ERR;
		}
		pam_syslog(pamh, loglevel, "set_loginuid failed opening loginuid");
		return rc;
	}
	if (pam_modutil_write(fd, loginuid, count) != count)
		rc = 1;
	close(fd);
	return rc;
}

#ifdef HAVE_LIBAUDIT
/*
 * This function is called only if "require_auditd" option is passed. It is
 * called after loginuid has been set. The purpose is to disallow logins
 * should the audit daemon not be running or crashed. It returns PAM_SUCCESS
 * if the audit daemon is running  and PAM_SESSION_ERR otherwise.
 */
static int check_auditd(void)
{
	int fd, retval;

	fd = audit_open();
	if (fd < 0) {
		/* This is here to let people that build their own kernel
		   and disable the audit system get in. You get these error
		   codes only when the kernel doesn't have audit
		   compiled in. */
		if (errno == EINVAL || errno == EPROTONOSUPPORT ||
		    errno == EAFNOSUPPORT)
			return PAM_SUCCESS;
		return PAM_SESSION_ERR;
	}
	retval = audit_request_status(fd);
	if (retval > 0) {
		struct audit_reply rep;
		int i;
		int timeout = 30; /* tenths of seconds */
		fd_set read_mask;

		FD_ZERO(&read_mask);
		FD_SET(fd, &read_mask);

		for (i = 0; i < timeout; i++) {
			struct timeval t;
			int rc;

			t.tv_sec  = 0;
			t.tv_usec = 100000;
			do {
				rc = select(fd+1, &read_mask, NULL, NULL, &t);
			} while (rc < 0 && errno == EINTR);

			rc = audit_get_reply(fd, &rep, GET_REPLY_NONBLOCKING,0);
			if (rc > 0) {
				/* If we get done or error, break out */
				if (rep.type == NLMSG_DONE ||
						rep.type == NLMSG_ERROR)
					break;

				/* If its not status, keep looping */
				if (rep.type != AUDIT_GET)
					continue;

				/* Found it... */
				close(fd);
				if (rep.status->pid == 0)
					return PAM_SESSION_ERR;
				else
					return PAM_SUCCESS;
			}
		}
	}
	close(fd);
	if (retval == -ECONNREFUSED) {
		/* This is here to let people that build their own kernel
		   and disable the audit system get in. ECONNREFUSED is
		   issued by the kernel when there is "no on listening". */
		return PAM_SUCCESS;
	} else if (retval == -EPERM && getuid() != 0) {
		/* If we get this, then the kernel supports auditing
		 * but we don't have enough privilege to write to the
		 * socket. Therefore, we have already been authenticated
		 * and we are a common user. Just act as though auditing
		 * is not enabled. Any other error we take seriously. */
		return PAM_SUCCESS;
	}

	return PAM_SESSION_ERR;
}
#endif

/*
 * Initialize audit session for user
 */
static int
_pam_loginuid(pam_handle_t *pamh, int flags UNUSED,
#ifdef HAVE_LIBAUDIT
	      int argc, const char **argv
#else
	      int argc UNUSED, const char **argv UNUSED
#endif
)
{
        const char *user = NULL;
	struct passwd *pwd;
#ifdef HAVE_LIBAUDIT
	int require_auditd = 0;
#endif

	/* get user name */
	if (pam_get_user(pamh, &user, NULL) != PAM_SUCCESS)
	{
		pam_syslog(pamh, LOG_ERR, "error recovering login user-name");
		return PAM_SESSION_ERR;
	}

        /* get user info */
	if ((pwd = pam_modutil_getpwnam(pamh, user)) == NULL) {
		pam_syslog(pamh, LOG_ERR,
			 "error: login user-name '%s' does not exist", user);
		return PAM_SESSION_ERR;
	}

	if (set_loginuid(pamh, pwd->pw_uid)) {
		pam_syslog(pamh, LOG_ERR, "set_loginuid failed\n");
		return PAM_SESSION_ERR;
	}

#ifdef HAVE_LIBAUDIT
	while (argc-- > 0) {
		if (strcmp(*argv, "require_auditd") == 0)
			require_auditd = 1;
		argv++;
	}

	if (require_auditd)
		return check_auditd();
	else
#endif
		return PAM_SUCCESS;
}

/*
 * PAM routines
 *
 * This is here for vsftpd which doesn't seem to run the session stack
 */
PAM_EXTERN int
pam_sm_acct_mgmt(pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	return _pam_loginuid(pamh, flags, argc, argv);
}

PAM_EXTERN int
pam_sm_open_session(pam_handle_t *pamh, int flags, int argc, const char **argv)
{
	return _pam_loginuid(pamh, flags, argc, argv);
}

PAM_EXTERN int
pam_sm_close_session(pam_handle_t *pamh UNUSED, int flags UNUSED,
		     int argc UNUSED, const char **argv UNUSED)
{
	return PAM_SUCCESS;
}

/* static module data */
#ifdef PAM_STATIC
struct pam_module _pam_loginuid_modstruct = {
    "pam_loginuid",
    NULL,
    NULL,
    pam_sm_acct_mgmt,
    pam_sm_open_session,
    pam_sm_close_session,
    NULL
};
#endif
